### PR TITLE
(enhancement) ItemVirtualMachine twig update

### PR DESCRIPTION
**Pull request by Lainow**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 

I need it for the print preview functionality because some tables are in twig and others are not, which makes the preview display complicated to modify effectively.

**Before**
image

**After**
image

### CODE REVIEW

File: templates/components/form/item_virtualmachine.html.twig [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{% if has_host %}
+    <h3 class='text-start'>{{ __('List of hosts') }}</h3>
+{% else %}
+    <h3 class='text-start mt-5'>{{ __('List of virtualized environments') }}</h3>
+    {% if canedit %}
+        <div class='center firstbloc'>
+            <a class='btn btn-primary' href='{{ vmformurl }}'>
```
Comment (cconard96) [inline templates/components/form/item_virtualmachine.html.twig:7]: Typically, I've been changing the list UI to show forms like this directly in the page with the list if they don't have a ton of fields.
If this remains as an anchor tag though, it needs `role="button"`.

File: templates/components/form/item_virtualmachine.html.twig [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+                {{ __('Add a virtual machine') }}
+            </a>
+        </div>
+    {% endif %}
+
+    {% if not has_vm %}
+        <div class='alert alert-info'>
```
Comment (cconard96) [inline templates/components/form/item_virtualmachine.html.twig:14]: Should use the macro from `templates/components/alerts_macros.html.twig` to standardize the display of alerts.

**Review by cconard96 — CHANGES_REQUESTED** [review]

File: templates/components/form/item_virtualmachine.html.twig [hunk 3]
```diff
@@ -1,19 +1,52 @@
...
+{% import 'components/alerts_macros.html.twig' as alerts %}
+
 {% if has_host %}
     <h3 class='text-start'>{{ __('List of hosts') }}</h3>
 {% else %}
     <h3 class='text-start mt-5'>{{ __('List of virtualized environments') }}</h3>
     {% if canedit %}
         <div class='center firstbloc'>
-            <a class='btn btn-primary' href='{{ vmformurl }}'>
+            <a class='btn btn-primary' href='{{ vmformurl }}' >
```
Comment (cconard96) [inline templates/components/form/item_virtualmachine.html.twig:40]: Adding `role="button"` is still relevant for accessibility.

**Review by cconard96 — APPROVED** [review]

**Review by cconard96 — APPROVED** [review]

File: src/ItemVirtualMachine.php [hunk 4]
```diff
@@ -203,45 +201,51 @@ public static function showForVirtualMachine(CommonDBTM $asset)
...
-                        echo $computer->fields['name'];
+                        $columns = [
+                            'name' => __('Name'),
+                            'entity' => _n('Entity', 'Entities', 1)
+                        ];
+                        $entries[] = [
+                            'name' => $computer->fields['name'],
+                            'entity' => Dropdown::getDropdownName('glpi_entities', $computer->fields['entities_id'])
+                        ];
                     }
```
Comment (cedric-anne) [inline src/ItemVirtualMachine.php:232]: I guess `$computer->can($host['items_id'], READ)` may return a different result for each row. In this case, you will have a mix of entries with 4 columns and entries with 2 columns. Also, `$columns` is redefined on each loop iteration.

IMHO, `$columns` should be defined once, and then every entries should contain the 4 columns (with empty values if they should ot be readable).

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Review by AdrienClairembault — APPROVED** [review]
Ok for me with suggested changes from @cedric-anne.

File: src/ItemVirtualMachine.php [hunk 5]
```diff
@@ -203,45 +201,48 @@ public static function showForVirtualMachine(CommonDBTM $asset)
...
-                echo  "<tr class='noHover'><th colspan='2' >" . __('List of hosts') . "</th></tr>";
-
-                $header = "<tr><th>" . __('Name') . "</th>";
-                $header .= "<th>" . Entity::getTypeName(1) . "</th>";
-                $header .= "</tr>";
-                echo $header;
-
+                TemplateRenderer::getInstance()->display('components/form/item_virtualmachine.html.twig', [
+                    'has_host' => true,
+                ]);
```
Comment (cedric-anne) [inline src/ItemVirtualMachine.php:206]: This is not mandatory to use the template rendering machine to render a single line. Also, your template corresponds to 2 distincts cases that have no parameters in common. This kind of cases should be handled by 2 distinc template files.

```suggestion
                echo '<h3 class="text-start">' . __s('List of hosts') . '</h3>';
```

File: src/ItemVirtualMachine.php [hunk 6]
```diff
@@ -203,45 +201,48 @@ public static function showForVirtualMachine(CommonDBTM $asset)
...
+                        $entries[] = [
+                            'name' => $computer->getLink($host['items_id']),
+                            'serial' => $computer->fields['serial'],
+                            'comment' => $computer->fields['comment'],
+                            'entity' => Dropdown::getDropdownName('glpi_entities', $computer->fields['entities_id'])
+                        ];
                     } else {
-                        echo $computer->fields['name'];
+                        $entries[] = [
+                            'name' => $computer->fields['name'],
```
Comment (cedric-anne) [inline src/ItemVirtualMachine.php:219]: Should be espaced here due to the usage of the `raw_html` formatter.
```suggestion
                            'name' => htmlspecialchars($computer->fields['name']),
```

File: src/ItemVirtualMachine.php [hunk 7]
```diff
@@ -282,98 +274,69 @@ public static function showForAsset(CommonDBTM $asset)
...
-            $header .= "<th>" . __('Machine') . "</th>";
-            $header .= "</tr>";
-            echo $header;
-
-            $vm = new self();
+        if ($has_vm) {
+            $entries = [];
             foreach ($virtualmachines as $virtualmachine) {
+                $vm = new self();
                 $vm->getFromDB($virtualmachine['id']);
```
Comment (cedric-anne) [inline src/ItemVirtualMachine.php:292]:
```suggestion
                if (!$vm->getFromDB($virtualmachine['id'])) {
                    continue:
                }
```

File: src/ItemVirtualMachine.php [hunk 8]
```diff
@@ -282,98 +274,69 @@ public static function showForAsset(CommonDBTM $asset)
...
-                echo "</tr>";
-                Session::addToNavigateListItems('ItemVirtualMachine', $virtualmachine['id']);
+
+                $type = VirtualMachineType::getById($virtualmachine['virtualmachinetypes_id']);
+                $system = VirtualMachineSystem::getById($virtualmachine['virtualmachinesystems_id']);
+                $state = VirtualMachineState::getById($virtualmachine['virtualmachinestates_id']);
+
+                $entries[] = [
+                    'name'                      => $vm->getLink() ?? NOT_AVAILABLE,
+                    'comment'                   => $virtualmachine['comment'] ?? NOT_AVAILABLE,
```
Comment (cedric-anne) [inline src/ItemVirtualMachine.php:300]: There is no reason to have null/undefined values here.
```suggestion
                    'name'                      => $vm->getLink(),
                    'comment'                   => $virtualmachine['comment'],
```

File: src/ItemVirtualMachine.php [hunk 9]
```diff
@@ -282,98 +274,69 @@ public static function showForAsset(CommonDBTM $asset)
...
+                    'dynamic'                   => $virtualmachine['is_dynamic'] ? __('Yes') : __('No'),
+                    'virtualmachinesystems_id'  => $system ? $system->getLink() : NOT_AVAILABLE,
+                    'virtualmachinestates_id'   => $state ? $state->getLink() : NOT_AVAILABLE,
+                    'uuid'                      => $virtualmachine['uuid'] ?? NOT_AVAILABLE,
+                    'vcpu'                      => $virtualmachine['vcpu'] ?? NOT_AVAILABLE,
+                    'ram'                       => $virtualmachine['ram'] ?? NOT_AVAILABLE,
```
Comment (cedric-anne) [inline src/ItemVirtualMachine.php:306]:
```suggestion
                    'uuid'                      => $virtualmachine['uuid'],
                    'vcpu'                      => $virtualmachine['vcpu'],
                    'ram'                       => $virtualmachine['ram'],
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]
Should `NOT_AVAILABLE` be displayed instead of empty values ?

**Comment by cconard96** [comment]
> Should `NOT_AVAILABLE` be displayed instead of empty values ?

Maybe that needs clarified/standardized. In some cases, we leave empty values and other times use N/A. I'm not really sure what would be best. Maybe empty when the type of value is a number/string/etc but N/A when it is a link to another item/object?

File: src/ItemVirtualMachine.php [hunk 10]
```diff
@@ -282,98 +272,71 @@ public static function showForAsset(CommonDBTM $asset)
...
-                        }
-
-                        $url .= "&nbsp; " . Html::showToolTip($tooltip, ['display' => false]);
-                    } else {
-                        $url = $asset->fields['name'];
-                    }
-                    echo $url;
+                $vm = new self();
+                if (!$vm->getFromDB($virtualmachine['id'])) {
+                    continue:
```
Comment (cedric-anne) [inline src/ItemVirtualMachine.php:291]:
```suggestion
                    continue;
```

File: templates/components/form/item_virtualmachine.html.twig [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+ # You should have received a copy of the GNU General Public License
+ # along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ #
+ # ---------------------------------------------------------------------
+ #}
+{% import 'components/alerts_macros.html.twig' as alerts %}
+
+{% if has_host %}
+    <h3 class='text-start'>{{ __('List of hosts') }}</h3>
+{% else %}
```
Comment (cedric-anne) [inline templates/components/form/item_virtualmachine.html.twig:36]: This case is not needed anymore.

**Review by cedric-anne — CHANGES_REQUESTED** [review]

File: templates/components/form/item_virtualmachine.html.twig [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+            </a>
+        </div>
+    {% endif %}
+
+    {% if not has_vm %}
+        {{ alerts.alert_info(
+            __('No virtualized environment associated with the item')
+        ) }}
+    {% endif %}
+{% endif %}
```
Comment (cedric-anne) [inline templates/components/form/item_virtualmachine.html.twig:51]:
```suggestion
<h3 class='text-start mt-5'>{{ __('List of virtualized environments') }}</h3>
{% if canedit %}
    <div class='text-center mb-3'>
        <a class='btn btn-primary' href='{{ vmformurl }}' role="button">
            {{ __('Add a virtual machine') }}
        </a>
    </div>
{% endif %}

{% if not has_vm %}
    {{ alerts.alert_info(
        __('No virtualized environment associated with the item')
    ) }}
{% endif %}
```

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**